File: dotnet/cs/tests/SamplingConfig.cs
/*
 * File: SamplingConfig.cs
 * Project: tests
 * Created Date: 25/11/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 01/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */


namespace tests;

public class SamplingConfigTest
{
    private const uint SamplingFreqDivMin = 512;
    private const uint SamplingFreqDivMax = 0xFFFFFFFF;
    private const double FreqMin = (double)AUTD3.FPGAClkFreq / SamplingFreqDivMax;
    private const double FreqMax = (double)AUTD3.FPGAClkFreq / SamplingFreqDivMin;
    private const ulong PeriodMin = (ulong)(1000000000.0 / AUTD3.FPGAClkFreq * SamplingFreqDivMin);
    private const ulong PeriodMax = (ulong)(1000000000.0 / AUTD3.FPGAClkFreq * SamplingFreqDivMax);

    [Fact]
    public void SamplingConfigFromFreqDiv()
    {
        var config = SamplingConfiguration.FromFrequencyDivision(SamplingFreqDivMin);
        Assert.Equal(512u, config.FrequencyDivision);
        Assert.Equal(40e3, config.Frequency);
        Assert.Equal(TimeSpan.FromMicroseconds(25), config.Period);

        Assert.Throws<AUTDException>(() => _ = SamplingConfiguration.FromFrequencyDivision(SamplingFreqDivMin - 1));
    }

    [Fact]
    public void SamplingConfigFromFreq()
    {
        var config = SamplingConfiguration.FromFrequency(40e3);
        Assert.Equal(512u, config.FrequencyDivision);
        Assert.Equal(40e3, config.Frequency);
        Assert.Equal(TimeSpan.FromMicroseconds(25), config.Period);

        Assert.Throws<AUTDException>(() => _ = SamplingConfiguration.FromFrequency(FreqMin - 0.1));
        Assert.Throws<AUTDException>(() => _ = SamplingConfiguration.FromFrequency(FreqMax + 0.1));
    }

    [Fact]
    public void SamplingConfigFromPeriod()
    {
        var config = SamplingConfiguration.FromPeriod(TimeSpan.FromMicroseconds(25));
        Assert.Equal(512u, config.FrequencyDivision);
        Assert.Equal(40e3, config.Frequency);
        Assert.Equal(TimeSpan.FromMicroseconds(25), config.Period);

        Assert.Throws<AUTDException>(() => _ = SamplingConfiguration.FromPeriod(TimeSpan.FromMicroseconds(PeriodMin / 1000.0 - 1.0)));
        Assert.Throws<AUTDException>(() => _ = SamplingConfiguration.FromPeriod(TimeSpan.FromMicroseconds(PeriodMax / 1000.0 + 1.0)));
    }
}

File: cpp/tests/internal/sampling_config.cpp
// File: sampling_config.cpp
// Project: internal
// Created Date: 25/11/2023
// Author: Shun Suzuki
// -----
// Last Modified: 01/12/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
//

#include <gtest/gtest.h>

#include <autd3/internal/sampling_config.hpp>
#include <chrono>
#include <numbers>

static constexpr uint32_t SAMPLING_FREQ_DIV_MIN = 512;
static constexpr uint32_t SAMPLING_FREQ_DIV_MAX = 0xFFFFFFFF;
static constexpr double FREQ_MIN = static_cast<double>(autd3::internal::native_methods::FPGA_CLK_FREQ) / SAMPLING_FREQ_DIV_MAX;
static constexpr double FREQ_MAX = static_cast<double>(autd3::internal::native_methods::FPGA_CLK_FREQ) / SAMPLING_FREQ_DIV_MIN;
static constexpr uint64_t PERIOD_MIN = static_cast<uint64_t>(1000000000.0 / autd3::internal::native_methods::FPGA_CLK_FREQ * SAMPLING_FREQ_DIV_MIN);
static constexpr uint64_t PERIOD_MAX = static_cast<uint64_t>(1000000000.0 / autd3::internal::native_methods::FPGA_CLK_FREQ * SAMPLING_FREQ_DIV_MAX);

TEST(Internal, SamplingConfigWithFreqDiv) {
  const auto config = autd3::internal::SamplingConfiguration::from_frequency_division(512);
  ASSERT_EQ(config.frequency_division(), 512);
  ASSERT_EQ(config.frequency(), 40e3);
  ASSERT_EQ(config.period(), std::chrono::microseconds(25));

  ASSERT_THROW((void)autd3::internal::SamplingConfiguration::from_frequency_division(SAMPLING_FREQ_DIV_MIN - 1), autd3::internal::AUTDException);
}

TEST(Internal, SamplingConfigWithFreq) {
  const auto config = autd3::internal::SamplingConfiguration::from_frequency(40e3);
  ASSERT_EQ(config.frequency_division(), 512);
  ASSERT_EQ(config.frequency(), 40e3);
  ASSERT_EQ(config.period(), std::chrono::microseconds(25));

  ASSERT_THROW((void)autd3::internal::SamplingConfiguration::from_frequency(FREQ_MIN - 0.1), autd3::internal::AUTDException);
  ASSERT_THROW((void)autd3::internal::SamplingConfiguration::from_frequency(FREQ_MAX + 0.1), autd3::internal::AUTDException);
}

TEST(Internal, SamplingConfigWithPeriod) {
  const auto config = autd3::internal::SamplingConfiguration::from_period(std::chrono::microseconds(25));
  ASSERT_EQ(config.frequency_division(), 512);
  ASSERT_EQ(config.frequency(), 40e3);
  ASSERT_EQ(config.period(), std::chrono::microseconds(25));

  ASSERT_THROW((void)autd3::internal::SamplingConfiguration::from_period(std::chrono::nanoseconds(PERIOD_MIN - 1)),
               autd3::internal::AUTDException);
  ASSERT_THROW((void)autd3::internal::SamplingConfiguration::from_period(std::chrono::nanoseconds(PERIOD_MAX + 1)),
               autd3::internal::AUTDException);
}
